doc/48-Database-Migration-Hook.md

```diff
@@ -0,0 +1,267 @@
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+
+CREATE TABLE training_schema (
+  id int unsigned NOT NULL AUTO_INCREMENT,
+  version varchar(64) NOT NULL,
+  timestamp bigint unsigned NOT NULL,
+  success enum ('n', 'y') DEFAULT NULL,
+  reason text DEFAULT NULL,
+
+  PRIMARY KEY (id),
+  CONSTRAINT idx_training_schema_version UNIQUE (version)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC;
+
+INSERT INTO training_schema (version, timestamp, success)
+  VALUES ('1.0.0', UNIX_TIMESTAMP() * 1000, 'y');
+```
+
+Users of the module can now create and initialize the database:
+
+```
+CREATE DATABASE training;
+mysql training < /usr/share/icingaweb2/modules/training/schema/mysql.schema.sql
+```
+
+The table `training_schema` is used to track the migrations.
+Each migration will generate a row in this table to determine the state of the schema:
+
+```
+mysql> SELECT * FROM training_schema;
++----+---------+---------------+---------+--------+
+| id | version | timestamp     | success | reason |
++----+---------+---------------+---------+--------+
+|  1 | 1.0.0   | 1714488742000 | y       | NULL   |
++----+---------+---------------+---------+--------+
+```
+
+A migration could for example, add new tables, add new columns on tables or update existing columns.
+
+In the following section we will create a migration to update the `asset` table.
+
+## Database Migrations
+
+In order to use the database migration hook it's required to place the migrations scripts in the following directories:
+
+```
+mkdir schema/mysql-upgrades/
+mkdir schema/pgsql-upgrades/
+```
+
+The name of these files should be the version of the migration (e.g. `1.0.0.sql`, `1.1.0.sql`).
+
+Now we can create SQL files that update the existing schema.
+For example, we can adjust the length of the `serial_no` field:
+
+```
+vi schema/mysql-upgrades/1.1.0.sql
+ALTER TABLE asset MODIFY COLUMN serial_no varchar(128) NOT NULL;
+
+INSERT INTO training_schema (version, timestamp, success)
+  VALUES ('1.1.0', UNIX_TIMESTAMP() * 1000, 'y');
+```
+
+Each migration will also insert itself into the table that tracks the migrations.
+
+Every time we require an update to the database schema, we create a new migration file with a higher version.
+
+```
+ls -l  schema/mysql-upgrades/
+1.0.0.sql
+1.1.0.sql
+1.2.0.sql
+1.3.0.sql
+```
+
+Now we need to implement the `DbMigrationHook` Hook.
+
+First, we need to create a Model for the `training_schema` rows so that we can interact with the
+table in an object-oriented way.
+
+```php
+vi library/Training/Model/Schema.php
+
+<?php
+
+namespace Icinga\Module\Training\Model;
+
+use DateTime;
+use ipl\Orm\Behavior\BoolCast;
+use ipl\Orm\Behavior\MillisecondTimestamp;
+use ipl\Orm\Behaviors;
+use ipl\Orm\Model;
+
+/**
+ * A database model for Training schema version table
+ *
+ * @property int $id Unique identifier of the database schema entries
+ * @property string $version The current schema version of Icinga Web
```

Comment:
of ~Icinga Web~ this/the module

Reply:
Otherwise, LFTM!

Reply:
copy-paste  https://github.com/Icinga/icingaweb2-module-reporting/blob/main/library/Reporting/Model/Schema.php#L17

Reply:
```suggestion
 * @property string $version The current schema version of the module
```